### PR TITLE
[Lock] Expose an expiringDate and isExpired method in Lock

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #22452 
| License       | MIT
| Doc PR        | NA

This PR store the expiration date of the lock in the key and exposes public method to let the user know the state of the Lock expiration.

### CODE REVIEW

File: src/Symfony/Component/Lock/Key.php
Comment (maidmaid) [inline src/Symfony/Component/Lock/Key.php]: Why not throw an exception when ``$this->expiringDate <= $expiringDate``?

Comment (jderusse) [inline src/Symfony/Component/Lock/Key.php]: I would do that in a method was `setExpiringDate`. But I didn't want to backport such logic in the class who want to move foward the expiration date.

**Comment by fabpot** [comment]
What's the status of this PR? I'd like to merge/close all PRs about the lock component ASAP to avoid having to revert the component for 3.4.

**Comment by jderusse** [comment]
@fabpot PR waiting for review/approval. As far as I know, nothing to add from my side.

File: src/Symfony/Component/Lock/Key.php
Comment (fabpot) [inline src/Symfony/Component/Lock/Key.php]: The method name is not clear to me. `reduce` means that it's going to expire earlier than before, right. But you say that the purpose is to "move forward".

Comment (jderusse) [inline src/Symfony/Component/Lock/Key.php]: When several store are combined, they can define different expiration date, the goal is to take the shorter one. cf `if (.. &&  $this->expiringDate > $expiringDate) { set value}`

Indeed, reducing a date does not really make sens. What's about `reduceLifeLength` or `advanceExpiringDate`?

Comment (chalasr) [inline src/Symfony/Component/Lock/Key.php]: It's common to use `lifetime` or `ttl` when talking about storage (session, cache, authorizations, ...). Could we use the same wording here? (`reduceLifetime` could do the job then)

Comment (jderusse) [reply src/Symfony/Component/Lock/Key.php]: Finally replaced `reduceExpiringDate(\Datetime $newExpiringDate)` by `reduceLifetime(float $ttl)`

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @jderusse.

File: src/Symfony/Component/Lock/Key.php
Comment (xabbuh) [inline src/Symfony/Component/Lock/Key.php]: `\DateTimeImmutable|null`

File: src/Symfony/Component/Lock/Key.php
Comment (xabbuh) [inline src/Symfony/Component/Lock/Key.php]: I am not sure if it really is expected behaviour that the method silently ignores any argument if it is no smaller than the existing expiration date.

Comment (jderusse) [reply src/Symfony/Component/Lock/Key.php]: IMO it's not an issue if a store ask for a TTL greater than the current one.

**Pull request closed without merging**
